File: LeetCode/LeetCode.Net/Problems/Backtracking/Subsets.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCode.Problems.Backtracking;

public class FindSubsets
{
    private IList<IList<int>> Sets = new List<IList<int>>();

    public IList<IList<int>> Subsets(int[] nums)
    {
        Backtracking(nums, 0, new List<int>());

        return Sets.ToList();
    }

    private void Backtracking(int[] nums, int currentIndex, List<int> path)
    {
        Sets.Add(new List<int>(path));

        for (var i = currentIndex; i < nums.Length; i++)
        {
            path.Add(nums[i]);
            Backtracking(nums, i + 1, path);
            path.RemoveAt(path.Count - 1);
        }
    }

    // private void Backtracking(int[] nums, int currentIndex, List<int> path)
    // {
    //     var copy = new List<int>(path);
    //     Sets.Add(copy);

    //     for(var i = currentIndex; i < nums.Length; i++){
    //         Backtracking(nums, i+1, copy);
    //         copy.Add(nums[i]);
    //     }
    // }
}

File: DSA/Python/Backtracking/subsets.py


def subsets(nums: List[int]) -> List[List[int]]:
    subsets = []
    def backtrack(nums: List[int], path: List[int], index: int):
        subsets.append(list(path))
        
        for i in range(index, len(nums)):
            path.append(nums[i])
            backtrack(nums, path, i + 1)
            del path[-1] 
    
    backtrack(nums, [], 0) 
    
    return subsets